File: frontend/src/types/penjualan.ts
/* ==========================================================
   DETAIL PENJUALAN
========================================================== */

export interface DetailPenjualan {
  id: number;

  barang_id: number;

  qty: number;

  harga_jual: number;

  subtotal: number;
}


/* ==========================================================
   PENJUALAN
========================================================== */

export interface Penjualan {
  id: number;

  no_faktur: string;

  tanggal: string;

  pelanggan_id: number | null;

  subtotal: number;

  diskon: number;

  pajak: number;

  grand_total: number;

  metode_bayar: string;

  dibayar: number;

  kembalian: number;

  status: string;

  keterangan: string | null;

  detail: DetailPenjualan[];
}

/* ==========================================================
   ITEM KERANJANG
========================================================== */

export interface PenjualanItem {

  barang_id: number;

  nama_barang: string;

  harga_jual: number;

  qty: number;

}

export interface PenjualanItemCreate {
  barang_id: number;

  qty: number;

  harga_jual: number;
}
/* ==========================================================
   CREATE
========================================================== */

export interface PenjualanCreate {
  pelanggan_id: number | null;

  diskon: number;

  pajak: number;

  metode_bayar: string;

  dibayar: number;

  keterangan: string;

  items: PenjualanItemCreate[];
}


/* ==========================================================
   UPDATE
========================================================== */

export interface PenjualanUpdate {
  pelanggan_id?: number | null;

  diskon?: number;

  pajak?: number;

  metode_bayar?: string;

  dibayar?: number;

  status?: string;

  keterangan?: string;
}


/* ==========================================================
   RESPONSE LIST
========================================================== */

export interface PenjualanListResponse {
  data: Penjualan[];

  total: number;
}


/* ==========================================================
   METODE PEMBAYARAN
========================================================== */

export const METODE_PEMBAYARAN = [
  "TUNAI",
  "QRIS",
  "TRANSFER",
  "DEBIT",
] as const;

export type MetodePembayaran =
  typeof METODE_PEMBAYARAN[number];


/* ==========================================================
   STATUS PENJUALAN
========================================================== */

export const STATUS_PENJUALAN = [
  "DRAFT",
  "SELESAI",
  "DIBATALKAN",
] as const;

export type StatusPenjualan =
  typeof STATUS_PENJUALAN[number];

File: backend/app/schemas/penjualan.py
from datetime import datetime
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


# ==========================================================
# DETAIL ITEM REQUEST
# ==========================================================

class PenjualanItem(BaseModel):

    barang_id: int = Field(
        ...,
        gt=0,
        examples=[16]
    )

    qty: int = Field(
        ...,
        gt=0,
        examples=[2]
    )

    harga_jual: Decimal = Field(
        ...,
        ge=0,
        examples=[15000]
    )


# ==========================================================
# CREATE REQUEST
# ==========================================================

class PenjualanCreate(BaseModel):

    pelanggan_id: Optional[int] = Field(
        default=None,
        examples=[1]
    )

    diskon: Decimal = Field(
        default=Decimal("0"),
        ge=0,
        examples=[0]
    )

    pajak: Decimal = Field(
        default=Decimal("0"),
        ge=0,
        examples=[0]
    )

    metode_bayar: str = Field(
        default="TUNAI",
        examples=["TUNAI"]
    )

    dibayar: Decimal = Field(
        default=Decimal("0"),
        ge=0,
        examples=[30000]
    )

    keterangan: Optional[str] = Field(
        default=None,
        examples=["Penjualan TEST"]
    )

    items: List[PenjualanItem] = Field(
        ...,
        min_length=1
    )


# ==========================================================
# UPDATE REQUEST
# ==========================================================

class PenjualanUpdate(BaseModel):

    pelanggan_id: Optional[int] = Field(
        default=None,
        examples=[1]
    )

    diskon: Optional[Decimal] = Field(
        default=None,
        ge=0,
        examples=[0]
    )

    pajak: Optional[Decimal] = Field(
        default=None,
        ge=0,
        examples=[0]
    )

    metode_bayar: Optional[str] = Field(
        default=None,
        examples=["TUNAI"]
    )

    dibayar: Optional[Decimal] = Field(
        default=None,
        ge=0,
        examples=[30000]
    )

    status: Optional[str] = Field(
        default=None,
        examples=["SELESAI"]
    )

    keterangan: Optional[str] = Field(
        default=None,
        examples=["Update penjualan"]
    )


# ==========================================================
# DETAIL RESPONSE
# ==========================================================

class DetailPenjualanResponse(BaseModel):

    id: int = Field(
        ...,
        examples=[1]
    )

    barang_id: int = Field(
        ...,
        examples=[16]
    )

    qty: int = Field(
        ...,
        examples=[2]
    )

    harga_jual: Decimal = Field(
        ...,
        examples=[15000]
    )

    subtotal: Decimal = Field(
        ...,
        examples=[30000]
    )

    model_config = ConfigDict(
        from_attributes=True
    )


# ==========================================================
# RESPONSE
# ==========================================================

class PenjualanResponse(BaseModel):

    id: int = Field(
        ...,
        examples=[21]
    )

    no_faktur: str = Field(
        ...,
        examples=["PJ-20260817-000001"]
    )

    tanggal: datetime

    pelanggan_id: Optional[int] = Field(
        default=None,
        examples=[1]
    )

    subtotal: Decimal = Field(
        ...,
        examples=[30000]
    )

    diskon: Decimal = Field(
        ...,
        examples=[0]
    )

    pajak: Decimal = Field(
        ...,
        examples=[0]
    )

    grand_total: Decimal = Field(
        ...,
        examples=[30000]
    )

    metode_bayar: str = Field(
        ...,
        examples=["TUNAI"]
    )

    dibayar: Decimal = Field(
        ...,
        examples=[30000]
    )

    kembalian: Decimal = Field(
        ...,
        examples=[0]
    )

    status: str = Field(
        ...,
        examples=["SELESAI"]
    )

    keterangan: Optional[str] = Field(
        default=None,
        examples=["Penjualan TEST FINAL KasirAI"]
    )

    detail: List[DetailPenjualanResponse] = Field(
        default_factory=list
    )

    model_config = ConfigDict(
        from_attributes=True
    )


# ==========================================================
# LIST RESPONSE
# ==========================================================

class PenjualanListResponse(BaseModel):

    data: List[PenjualanResponse]

    total: int = Field(
        ...,
        examples=[8]
    )

    model_config = ConfigDict(
        from_attributes=True
    )